File: vscode-extension/src/repoCommands.ts
import * as fs from "fs/promises";
import * as path from "path";
import { parse, ParseError, printParseErrorCode } from "jsonc-parser";
import { beginFilesystemRead } from "./performanceMetrics";

export const REPO_COMMAND_MANIFEST_PATH = path.join(
  "configs",
  "freecm.commands.jsonc",
);

export type RepoCommandAction = "config" | "build" | "run" | "test" | "package";
export type RepoCommandDependentAction = Exclude<RepoCommandAction, "config">;

export interface RepoCommandStep {
  readonly command: string;
  readonly args: readonly string[];
}

export interface RepoCommandReadinessSpec {
  readonly inputs: readonly string[];
  readonly outputs: readonly string[];
}

export interface RepoCommandVariant {
  readonly id: string;
  readonly label: string;
  readonly command?: string;
  readonly args?: readonly string[];
  readonly steps: readonly RepoCommandStep[];
  readonly description?: string;
  readonly platforms?: readonly string[];
  readonly default?: boolean;
  readonly defaults?: Readonly<
    Partial<Record<RepoCommandDependentAction, string>>
  >;
  readonly configurations?: readonly string[];
  readonly readiness?: RepoCommandReadinessSpec;
}

export interface RepoCommandActionState {
  readonly action: RepoCommandAction;
  readonly variants: readonly RepoCommandVariant[];
}

export interface RepoCommandManifestState {
  readonly manifestPath: string;
  readonly actions: Record<RepoCommandAction, RepoCommandActionState>;
  readonly configurations: readonly RepoCommandVariant[];
  readonly defaultConfiguration: RepoCommandVariant | undefined;
}

export interface RepoCommandWarning {
  readonly action: RepoCommandAction;
  readonly variantId: string;
  readonly variantLabel: string;
  readonly stepIndex: number;
  readonly message: string;
}

export const REPO_COMMAND_ACTIONS = [
  "config",
  "build",
  "run",
  "test",
  "package",
] as const;
export const REPO_COMMAND_DEPENDENT_ACTIONS = [
  "build",
  "run",
  "test",
  "package",
] as const satisfies readonly RepoCommandDependentAction[];
export const SUPPORTED_REPO_COMMAND_PLATFORMS = [
  "darwin",
  "linux",
  "win32",
] as const;

const SUPPORTED_VERSION = 2;

export async function loadRepoCommandManifest(
  repoRoot: string,
  platform: string = process.platform,
): Promise<RepoCommandManifestState | undefined> {
  const manifestPath = path.join(repoRoot, REPO_COMMAND_MANIFEST_PATH);
  let text: string;
  const finishRead = beginFilesystemRead();
  try {
    text = await fs.readFile(manifestPath, "utf8");
  } catch (error) {
    if (isNodeErrorCode(error, "ENOENT")) {
      return undefined;
    }
    throw new Error(`Unable to read ${manifestPath}: ${errorMessage(error)}`);
  } finally {
    finishRead();
  }

  return parseRepoCommandManifest(text, manifestPath, platform);
}

export function parseRepoCommandManifest(
  text: string,
  manifestPath: string,
  platform: string,
): RepoCommandManifestState {
  if (
    !SUPPORTED_REPO_COMMAND_PLATFORMS.includes(
      platform as (typeof SUPPORTED_REPO_COMMAND_PLATFORMS)[number],
    )
  ) {
    throw new Error(
      `Invalid command manifest platform ${JSON.stringify(
        platform,
      )}: expected darwin, linux, or win32`,
    );
  }

  const errors: ParseError[] = [];
  const value = parse(text, errors, { allowTrailingComma: true });
  if (errors.length > 0) {
    const details = errors
      .map(
        (error) =>
          `${printParseErrorCode(error.error)} at offset ${error.offset}`,
      )
      .join(", ");
    throw new Error(`Invalid JSONC in ${manifestPath}: ${details}`);
  }
  if (!isObject(value)) {
    throw new Error(
      `Invalid command manifest ${manifestPath}: expected top-level object`,
    );
  }
  if (value.version !== SUPPORTED_VERSION) {
    throw new Error(
      `Invalid command manifest ${manifestPath}: version must be ${SUPPORTED_VERSION}`,
    );
  }
  if (!isObject(value.commands)) {
    throw new Error(
      `Invalid command manifest ${manifestPath}: commands must be an object`,
    );
  }
  const commands = value.commands;

  const parsedActions = Object.fromEntries(
    REPO_COMMAND_ACTIONS.map((action) => [
      action,
      parseAction(action, commands[action], manifestPath),
    ]),
  ) as Record<RepoCommandAction, RepoCommandActionState>;
  validateManifestRelationships(parsedActions, manifestPath);

  const configurations = parsedActions.config.variants.filter((variant) =>
    isPlatformCompatible(variant, platform),
  );
  const platformConfigurationIds = new Set(
    configurations.map((configuration) => configuration.id),
  );
  const platformAction = (
    action: RepoCommandDependentAction,
  ): RepoCommandActionState => ({
    action,
    variants: parsedActions[action].variants.filter((variant) =>
      variant.configurations?.some((configurationId) =>
        platformConfigurationIds.has(configurationId),
      ),
    ),
  });
  const actions: Record<RepoCommandAction, RepoCommandActionState> = {
    config: { action: "config", variants: configurations },
    build: platformAction("build"),
    run: platformAction("run"),
    test: platformAction("test"),
    package: platformAction("package"),
  };

  return {
    manifestPath,
    actions,
    configurations,
    defaultConfiguration: configurations.find(
      (configuration) => configuration.default === true,
    ),
  };
}

function parseAction(
  action: RepoCommandAction,
  value: unknown,
  manifestPath: string,
): RepoCommandActionState {
  if (value === undefined) {
    return { action, variants: [] };
  }
  if (!Array.isArray(value)) {
    throw new Error(
      `Invalid command manifest ${manifestPath}: commands.${action} must be an array`,
    );
  }

  const variants = value.map((entry, index) =>
    parseVariant(action, entry, index, manifestPath),
  );
  assertUniqueVariantIds(action, variants, manifestPath);
  return { action, variants };
}

function parseVariant(
  action: RepoCommandAction,
  value: unknown,
  index: number,
  manifestPath: string,
): RepoCommandVariant {
  const prefix = `commands.${action}[${index}]`;
  if (!isObject(value)) {
    throw new Error(
      `Invalid command manifest ${manifestPath}: ${prefix} must be an object`,
    );
  }

  const id = requiredString(value.id, manifestPath, `${prefix}.id`);
  const label = requiredString(value.label, manifestPath, `${prefix}.label`);
  const steps = parseVariantSteps(value, manifestPath, prefix);
  const variant: {
    id: string;
    label: string;
    command?: string;
    args?: string[];
    steps: RepoCommandStep[];
    description?: string;
    platforms?: string[];
    default?: boolean;
    defaults?: Partial<Record<RepoCommandDependentAction, string>>;
    configurations?: string[];
    readiness?: RepoCommandReadinessSpec;
  } = {
    id,
    label,
    steps,
  };

  if (value.steps === undefined) {
    variant.command = steps[0].command;
    variant.args = [...steps[0].args];
  }
  if (value.description !== undefined) {
    variantWithOptionalString(
      value.description,
      manifestPath,
      `${prefix}.description`,
    );
    variant.description = value.description as string;
  }

  if (action === "config") {
    parseConfigVariantFields(value, variant, manifestPath, prefix);
  } else {
    parseDependentVariantFields(value, variant, manifestPath, prefix);
  }
  return variant;
}

function parseConfigVariantFields(
  value: Record<string, unknown>,
  variant: {
    platforms?: string[];
    default?: boolean;
    defaults?: Partial<Record<RepoCommandDependentAction, string>>;
    readiness?: RepoCommandReadinessSpec;
  },
  manifestPath: string,
  prefix: string,
): void {
  if (value.configurations !== undefined) {
    throw new Error(
      `Invalid command manifest ${manifestPath}: ${prefix}.configurations is only valid for downstream actions`,
    );
  }

  if (value.platforms !== undefined) {
    const platforms = requiredNonEmptyStringArray(
      value.platforms,
      manifestPath,
      `${prefix}.platforms`,
    );
    for (const platform of platforms) {
      if (
        !SUPPORTED_REPO_COMMAND_PLATFORMS.includes(
          platform as (typeof SUPPORTED_REPO_COMMAND_PLATFORMS)[number],
        )
      ) {
        throw new Error(
          `Invalid command manifest ${manifestPath}: ${prefix}.platforms contains unsupported platform ${JSON.stringify(
            platform,
          )}`,
        );
      }
    }
    assertUniqueStrings(platforms, manifestPath, `${prefix}.platforms`);
    variant.platforms = platforms;
  }

  if (value.default !== undefined) {
    if (typeof value.default !== "boolean") {
      throw new Error(
        `Invalid command manifest ${manifestPath}: ${prefix}.default must be boolean`,
      );
    }
    variant.default = value.default;
  }

  variant.defaults = parseConfigurationDefaults(
    value.defaults,
    manifestPath,
    `${prefix}.defaults`,
  );
  if (value.readiness !== undefined) {
    variant.readiness = parseReadinessSpec(
      value.readiness,
      manifestPath,
      `${prefix}.readiness`,
    );
  }
}

function parseDependentVariantFields(
  value: Record<string, unknown>,
  variant: { configurations?: string[] },
  manifestPath: string,
  prefix: string,
): void {
  for (const configOnlyField of [
    "platforms",
    "default",
    "defaults",
    "readiness",
  ] as const) {
    if (value[configOnlyField] !== undefined) {
      throw new Error(
        `Invalid command manifest ${manifestPath}: ${prefix}.${configOnlyField} is only valid for Config variants`,
      );
    }
  }

  const configurations = requiredNonEmptyStringArray(
    value.configurations,
    manifestPath,
    `${prefix}.configurations`,
  );
  assertUniqueStrings(
    configurations,
    manifestPath,
    `${prefix}.configurations`,
  );
  variant.configurations = configurations;
}

function parseConfigurationDefaults(
  value: unknown,
  manifestPath: string,
  fieldPath: string,
): Partial<Record<RepoCommandDependentAction, string>> {
  if (!isObject(value)) {
    throw new Error(
      `Invalid command manifest ${manifestPath}: ${fieldPath} must be an object`,
    );
  }

  const defaults: Partial<Record<RepoCommandDependentAction, string>> = {};
  for (const [action, variantId] of Object.entries(value)) {
    if (
      !REPO_COMMAND_DEPENDENT_ACTIONS.includes(
        action as RepoCommandDependentAction,
      )
    ) {
      throw new Error(
        `Invalid command manifest ${manifestPath}: ${fieldPath} contains unsupported action ${JSON.stringify(
          action,
        )}`,
      );
    }
    defaults[action as RepoCommandDependentAction] = requiredString(
      variantId,
      manifestPath,
      `${fieldPath}.${action}`,
    );
  }
  return defaults;
}

function parseReadinessSpec(
  value: unknown,
  manifestPath: string,
  fieldPath: string,
): RepoCommandReadinessSpec {
  if (!isObject(value)) {
    throw new Error(
      `Invalid command manifest ${manifestPath}: ${fieldPath} must be an object`,
    );
  }

  const inputs = optionalStringArray(
    value.inputs,
    manifestPath,
    `${fieldPath}.inputs`,
  );
  const outputs = optionalStringArray(
    value.outputs,
    manifestPath,
    `${fieldPath}.outputs`,
  );
  assertUniqueStrings(inputs, manifestPath, `${fieldPath}.inputs`);
  assertUniqueStrings(outputs, manifestPath, `${fieldPath}.outputs`);
  for (const [kind, values] of [
    ["inputs", inputs],
    ["outputs", outputs],
  ] as const) {
    for (const repoPath of values) {
      assertRepoRelativePath(
        repoPath,
        manifestPath,
        `${fieldPath}.${kind}`,
      );
    }
  }
  return { inputs, outputs };
}

export function compatibleRepoCommandVariants(
  manifest: RepoCommandManifestState,
  configurationId: string,
  action: RepoCommandDependentAction,
): readonly RepoCommandVariant[] {
  if (
    !manifest.configurations.some(
      (configuration) => configuration.id === configurationId,
    )
  ) {
    return [];
  }
  return manifest.actions[action].variants.filter((variant) =>
    variant.configurations?.includes(configurationId),
  );
}

export function defaultRepoCommandVariant(
  manifest: RepoCommandManifestState,
  configurationId: string,
  action: RepoCommandDependentAction,
): RepoCommandVariant | undefined {
  const configuration = manifest.configurations.find(
    (candidate) => candidate.id === configurationId,
  );
  const defaultId = configuration?.defaults?.[action];
  if (defaultId === undefined) {
    return undefined;
  }
  return compatibleRepoCommandVariants(
    manifest,
    configurationId,
    action,
  ).find((variant) => variant.id === defaultId);
}

export function isRepoCommandVariantCompatible(
  variant: RepoCommandVariant,
  configurationId: string,
): boolean {
  return variant.configurations?.includes(configurationId) === true;
}

export function commandLineForTerminal(
  variant: RepoCommandVariant,
  platform: string = process.platform,
): string {
  return commandLinesForTerminal(variant, platform).join("\n");
}

export function commandLinesForTerminal(
  variant: RepoCommandVariant,
  platform: string = process.platform,
): string[] {
  const quote = platform === "win32" ? windowsShellQuote : posixShellQuote;
  return variant.steps.map((step) =>
    [step.command, ...step.args].map(quote).join(" "),
  );
}

export function repoCommandWarnings(
  manifest: RepoCommandManifestState,
  platform: string = process.platform,
): RepoCommandWarning[] {
  return REPO_COMMAND_ACTIONS.flatMap((action) =>
    manifest.actions[action].variants.flatMap((variant) =>
      variant.steps.flatMap((step, index) =>
        repoCommandStepWarnings(action, variant, step, index, platform),
      ),
    ),
  );
}

function repoCommandStepWarnings(
  action: RepoCommandAction,
  variant: RepoCommandVariant,
  step: RepoCommandStep,
  stepIndex: number,
  platform: string,
): RepoCommandWarning[] {
  if (
    platform === "darwin" &&
    action === "run" &&
    step.command === "open" &&
    step.args.some((arg) => arg.endsWith(".app"))
  ) {
    return [
      {
        action,
        variantId: variant.id,
        variantLabel: variant.label,
        stepIndex,
        message:
          "macOS run command uses open on a .app bundle; this detaches from the terminal. Prefer .app/Contents/MacOS/<ExecutableName> so logs stay attached and Ctrl+C can stop the app.",
      },
    ];
  }
  return [];
}

function parseVariantSteps(
  value: Record<string, unknown>,
  manifestPath: string,
  prefix: string,
): RepoCommandStep[] {
  if (value.steps !== undefined) {
    if (value.command !== undefined || value.args !== undefined) {
      throw new Error(
        `Invalid command manifest ${manifestPath}: ${prefix} must use either command/args or steps, not both`,
      );
    }
    if (!Array.isArray(value.steps) || value.steps.length === 0) {
      throw new Error(
        `Invalid command manifest ${manifestPath}: ${prefix}.steps must be a non-empty array`,
      );
    }
    return value.steps.map((step, index) =>
      parseCommandStep(step, manifestPath, `${prefix}.steps[${index}]`),
    );
  }

  return [
    {
      command: requiredString(value.command, manifestPath, `${prefix}.command`),
      args: requiredStringArray(value.args, manifestPath, `${prefix}.args`),
    },
  ];
}

function parseCommandStep(
  value: unknown,
  manifestPath: string,
  prefix: string,
): RepoCommandStep {
  if (!isObject(value)) {
    throw new Error(
      `Invalid command manifest ${manifestPath}: ${prefix} must be an object`,
    );
  }
  return {
    command: requiredString(value.command, manifestPath, `${prefix}.command`),
    args: requiredStringArray(value.args, manifestPath, `${prefix}.args`),
  };
}

function posixShellQuote(value: string): string {
  if (value.length > 0 && /^[A-Za-z0-9_./:=@%+-]+$/.test(value)) {
    return value;
  }
  return `'${value.replace(/'/g, `'\\''`)}'`;
}

function windowsShellQuote(value: string): string {
  if (value.length > 0 && /^[A-Za-z0-9_./:=@%+\\-]+$/.test(value)) {
    return value;
  }
  return `"${value.replace(/"/g, '\\"')}"`;
}

function assertUniqueVariantIds(
  action: RepoCommandAction,
  variants: readonly RepoCommandVariant[],
  manifestPath: string,
): void {
  const seen = new Set<string>();
  for (const variant of variants) {
    if (seen.has(variant.id)) {
      throw new Error(
        `Invalid command manifest ${manifestPath}: commands.${action} contains duplicate id ${JSON.stringify(
          variant.id,
        )}`,
      );
    }
    seen.add(variant.id);
  }
}

function validateManifestRelationships(
  actions: Record<RepoCommandAction, RepoCommandActionState>,
  manifestPath: string,
): void {
  const configurations = actions.config.variants;
  const configurationIds = new Set(
    configurations.map((configuration) => configuration.id),
  );

  for (const action of REPO_COMMAND_DEPENDENT_ACTIONS) {
    for (const variant of actions[action].variants) {
      for (const configurationId of variant.configurations ?? []) {
        if (!configurationIds.has(configurationId)) {
          throw new Error(
            `Invalid command manifest ${manifestPath}: commands.${action} variant ${JSON.stringify(
              variant.id,
            )} references unknown Config ${JSON.stringify(configurationId)}`,
          );
        }
      }
    }
  }

  for (const configuration of configurations) {
    for (const action of REPO_COMMAND_DEPENDENT_ACTIONS) {
      const variants = actions[action].variants.filter((variant) =>
        variant.configurations?.includes(configuration.id),
      );
      const defaultId = configuration.defaults?.[action];
      if (variants.length === 0) {
        if (defaultId !== undefined) {
          throw new Error(
            `Invalid command manifest ${manifestPath}: Config ${JSON.stringify(
              configuration.id,
            )} defaults.${action} references ${JSON.stringify(
              defaultId,
            )}, but no ${action} variants support that Config`,
          );
        }
        continue;
      }
      if (defaultId === undefined) {
        throw new Error(
          `Invalid command manifest ${manifestPath}: Config ${JSON.stringify(
            configuration.id,
          )} must declare defaults.${action} because compatible ${action} variants exist`,
        );
      }
      if (!variants.some((variant) => variant.id === defaultId)) {
        throw new Error(
          `Invalid command manifest ${manifestPath}: Config ${JSON.stringify(
            configuration.id,
          )} defaults.${action} ${JSON.stringify(
            defaultId,
          )} is not a compatible ${action} variant`,
        );
      }
    }
  }

  for (const platform of SUPPORTED_REPO_COMMAND_PLATFORMS) {
    const compatibleConfigurations = configurations.filter((configuration) =>
      isPlatformCompatible(configuration, platform),
    );
    if (compatibleConfigurations.length === 0) {
      continue;
    }
    const defaults = compatibleConfigurations.filter(
      (configuration) => configuration.default === true,
    );
    if (defaults.length !== 1) {
      throw new Error(
        `Invalid command manifest ${manifestPath}: platform ${platform} must have exactly one default Config; found ${defaults.length}`,
      );
    }
  }
}

function requiredString(
  value: unknown,
  manifestPath: string,
  fieldPath: string,
): string {
  if (typeof value !== "string" || value.trim() === "") {
    throw new Error(
      `Invalid command manifest ${manifestPath}: ${fieldPath} must be a non-empty string`,
    );
  }
  return value;
}

function requiredStringArray(
  value: unknown,
  manifestPath: string,
  fieldPath: string,
): string[] {
  if (
    !Array.isArray(value) ||
    !value.every((entry) => typeof entry === "string")
  ) {
    throw new Error(
      `Invalid command manifest ${manifestPath}: ${fieldPath} must be a string array`,
    );
  }
  return [...value];
}

function requiredNonEmptyStringArray(
  value: unknown,
  manifestPath: string,
  fieldPath: string,
): string[] {
  const result = requiredStringArray(value, manifestPath, fieldPath);
  if (result.length === 0 || result.some((entry) => entry.trim() === "")) {
    throw new Error(
      `Invalid command manifest ${manifestPath}: ${fieldPath} must be a non-empty string array`,
    );
  }
  return result;
}

function optionalStringArray(
  value: unknown,
  manifestPath: string,
  fieldPath: string,
): string[] {
  if (value === undefined) {
    return [];
  }
  const result = requiredStringArray(value, manifestPath, fieldPath);
  if (result.some((entry) => entry.trim() === "")) {
    throw new Error(
      `Invalid command manifest ${manifestPath}: ${fieldPath} entries must be non-empty strings`,
    );
  }
  return result;
}

function assertUniqueStrings(
  values: readonly string[],
  manifestPath: string,
  fieldPath: string,
): void {
  const seen = new Set<string>();
  for (const value of values) {
    if (seen.has(value)) {
      throw new Error(
        `Invalid command manifest ${manifestPath}: ${fieldPath} contains duplicate value ${JSON.stringify(
          value,
        )}`,
      );
    }
    seen.add(value);
  }
}

function assertRepoRelativePath(
  value: string,
  manifestPath: string,
  fieldPath: string,
): void {
  const parts = value.split(/[\\/]/);
  if (
    path.isAbsolute(value) ||
    path.win32.isAbsolute(value) ||
    parts.includes("..")
  ) {
    throw new Error(
      `Invalid command manifest ${manifestPath}: ${fieldPath} path ${JSON.stringify(
        value,
      )} must stay within the repository`,
    );
  }
}

function variantWithOptionalString(
  value: unknown,
  manifestPath: string,
  fieldPath: string,
): void {
  if (typeof value !== "string") {
    throw new Error(
      `Invalid command manifest ${manifestPath}: ${fieldPath} must be a string`,
    );
  }
}

function isPlatformCompatible(
  variant: RepoCommandVariant,
  platform: string,
): boolean {
  return (
    variant.platforms === undefined || variant.platforms.includes(platform)
  );
}

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function isNodeErrorCode(error: unknown, code: string): boolean {
  return isObject(error) && error.code === code;
}

function errorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

File: freecm/repo_commands.py
"""Lightweight validation for downstream repository command manifests."""

from __future__ import annotations

import json
from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from pathlib import Path, PurePosixPath, PureWindowsPath
from typing import Any, NoReturn

from .errors import RepoCommandManifestError
from .jsonc import strip_jsonc_comments, strip_jsonc_trailing_commas

REPO_COMMAND_MANIFEST_PATH = Path("configs/freecm.commands.jsonc")
REPO_COMMAND_ACTIONS = ("config", "build", "run", "test", "package")
REPO_COMMAND_DEPENDENT_ACTIONS = ("build", "run", "test", "package")
SUPPORTED_REPO_COMMAND_PLATFORMS = ("darwin", "linux", "win32")
SUPPORTED_REPO_COMMAND_MANIFEST_VERSION = 2
_MISSING = object()


@dataclass(frozen=True)
class RepoCommandManifestSummary:
    """Small success receipt returned by the headless validator."""

    manifest_path: Path
    configuration_count: int
    variant_count: int


@dataclass(frozen=True)
class _Variant:
    id: str
    platforms: tuple[str, ...] | None = None
    default: bool = False
    defaults: Mapping[str, str] | None = None
    configurations: tuple[str, ...] = ()


def validate_repo_command_manifest(repo_root: str | Path) -> RepoCommandManifestSummary:
    """Validate one downstream manifest with no Node or third-party Python dependency."""

    manifest_path = Path(repo_root).resolve() / REPO_COMMAND_MANIFEST_PATH
    try:
        text = manifest_path.read_text(encoding="utf-8")
    except OSError as error:
        raise RepoCommandManifestError(f"Unable to read {manifest_path}: {error}") from error
    return validate_repo_command_manifest_text(text, manifest_path=manifest_path)


def validate_repo_command_manifest_text(
    text: str,
    *,
    manifest_path: str | Path = REPO_COMMAND_MANIFEST_PATH,
) -> RepoCommandManifestSummary:
    """Validate manifest text and return only a compact structural summary."""

    path = Path(manifest_path)
    value = _parse_jsonc(text, path)
    if not isinstance(value, dict):
        _fail(path, "expected top-level object")
    if value.get("version") != SUPPORTED_REPO_COMMAND_MANIFEST_VERSION:
        _fail(path, f"version must be {SUPPORTED_REPO_COMMAND_MANIFEST_VERSION}")

    commands = value.get("commands")
    if not isinstance(commands, dict):
        _fail(path, "commands must be an object")

    actions = {
        action: _parse_action(
            action,
            commands[action] if action in commands else _MISSING,
            path,
        )
        for action in REPO_COMMAND_ACTIONS
    }
    _validate_relationships(actions, path)
    return RepoCommandManifestSummary(
        manifest_path=path,
        configuration_count=len(actions["config"]),
        variant_count=sum(len(variants) for variants in actions.values()),
    )


def _parse_jsonc(text: str, manifest_path: Path) -> Any:
    normalized = strip_jsonc_trailing_commas(strip_jsonc_comments(text))
    try:
        return json.loads(normalized)
    except json.JSONDecodeError as error:
        raise RepoCommandManifestError(
            f"Invalid JSONC in {manifest_path}: {error.msg} "
            f"at line {error.lineno} column {error.colno}"
        ) from error


def _parse_action(action: str, value: object, manifest_path: Path) -> list[_Variant]:
    if value is _MISSING:
        return []
    if not isinstance(value, list):
        _fail(manifest_path, f"commands.{action} must be an array")

    variants = [
        _parse_variant(action, entry, index, manifest_path) for index, entry in enumerate(value)
    ]
    _assert_unique_strings(
        [variant.id for variant in variants],
        manifest_path,
        f"commands.{action}",
        noun="id",
    )
    return variants


def _parse_variant(
    action: str,
    value: object,
    index: int,
    manifest_path: Path,
) -> _Variant:
    prefix = f"commands.{action}[{index}]"
    if not isinstance(value, dict):
        _fail(manifest_path, f"{prefix} must be an object")

    variant_id = _required_string(value.get("id"), manifest_path, f"{prefix}.id")
    _required_string(value.get("label"), manifest_path, f"{prefix}.label")
    _parse_variant_steps(value, manifest_path, prefix)
    if "description" in value and not isinstance(value["description"], str):
        _fail(manifest_path, f"{prefix}.description must be a string")

    if action == "config":
        return _parse_config_fields(value, variant_id, manifest_path, prefix)
    return _parse_dependent_fields(value, variant_id, manifest_path, prefix)


def _parse_config_fields(
    value: Mapping[str, object],
    variant_id: str,
    manifest_path: Path,
    prefix: str,
) -> _Variant:
    if "configurations" in value:
        _fail(
            manifest_path,
            f"{prefix}.configurations is only valid for downstream actions",
        )

    platforms: tuple[str, ...] | None = None
    if "platforms" in value:
        platforms = _required_non_empty_string_array(
            value["platforms"], manifest_path, f"{prefix}.platforms"
        )
        _assert_unique_strings(platforms, manifest_path, f"{prefix}.platforms")
        for platform in platforms:
            if platform not in SUPPORTED_REPO_COMMAND_PLATFORMS:
                _fail(
                    manifest_path,
                    f"{prefix}.platforms contains unsupported platform {platform!r}",
                )

    default = value.get("default", False)
    if "default" in value and not isinstance(default, bool):
        _fail(manifest_path, f"{prefix}.default must be boolean")

    defaults_value = value.get("defaults")
    if not isinstance(defaults_value, dict):
        _fail(manifest_path, f"{prefix}.defaults must be an object")
    defaults: dict[str, str] = {}
    for action, referenced_id in defaults_value.items():
        if action not in REPO_COMMAND_DEPENDENT_ACTIONS:
            _fail(
                manifest_path,
                f"{prefix}.defaults contains unsupported action {action!r}",
            )
        defaults[action] = _required_string(
            referenced_id, manifest_path, f"{prefix}.defaults.{action}"
        )

    if "readiness" in value:
        _parse_readiness(value["readiness"], manifest_path, f"{prefix}.readiness")

    return _Variant(
        id=variant_id,
        platforms=platforms,
        default=bool(default),
        defaults=defaults,
    )


def _parse_dependent_fields(
    value: Mapping[str, object],
    variant_id: str,
    manifest_path: Path,
    prefix: str,
) -> _Variant:
    for config_only_field in ("platforms", "default", "defaults", "readiness"):
        if config_only_field in value:
            _fail(
                manifest_path,
                f"{prefix}.{config_only_field} is only valid for Config variants",
            )

    configurations = _required_non_empty_string_array(
        value.get("configurations"), manifest_path, f"{prefix}.configurations"
    )
    _assert_unique_strings(configurations, manifest_path, f"{prefix}.configurations")
    return _Variant(id=variant_id, configurations=configurations)


def _parse_readiness(value: object, manifest_path: Path, field_path: str) -> None:
    if not isinstance(value, dict):
        _fail(manifest_path, f"{field_path} must be an object")
    for kind in ("inputs", "outputs"):
        raw_paths = value[kind] if kind in value else _MISSING
        paths = _optional_string_array(raw_paths, manifest_path, f"{field_path}.{kind}")
        _assert_unique_strings(paths, manifest_path, f"{field_path}.{kind}")
        for repo_path in paths:
            _assert_repo_relative_path(repo_path, manifest_path, f"{field_path}.{kind}")


def _parse_variant_steps(
    value: Mapping[str, object],
    manifest_path: Path,
    prefix: str,
) -> None:
    if "steps" in value:
        if "command" in value or "args" in value:
            _fail(
                manifest_path,
                f"{prefix} must use either command/args or steps, not both",
            )
        steps = value["steps"]
        if not isinstance(steps, list) or not steps:
            _fail(manifest_path, f"{prefix}.steps must be a non-empty array")
        for index, step in enumerate(steps):
            _parse_command_step(step, manifest_path, f"{prefix}.steps[{index}]")
        return

    _required_string(value.get("command"), manifest_path, f"{prefix}.command")
    _required_string_array(value.get("args"), manifest_path, f"{prefix}.args")


def _parse_command_step(value: object, manifest_path: Path, prefix: str) -> None:
    if not isinstance(value, dict):
        _fail(manifest_path, f"{prefix} must be an object")
    _required_string(value.get("command"), manifest_path, f"{prefix}.command")
    _required_string_array(value.get("args"), manifest_path, f"{prefix}.args")


def _validate_relationships(
    actions: Mapping[str, Sequence[_Variant]],
    manifest_path: Path,
) -> None:
    configurations = actions["config"]
    configuration_ids = {configuration.id for configuration in configurations}

    for action in REPO_COMMAND_DEPENDENT_ACTIONS:
        for variant in actions[action]:
            for configuration_id in variant.configurations:
                if configuration_id not in configuration_ids:
                    _fail(
                        manifest_path,
                        f"commands.{action} variant {variant.id!r} references unknown "
                        f"Config {configuration_id!r}",
                    )

    for configuration in configurations:
        defaults = configuration.defaults or {}
        for action in REPO_COMMAND_DEPENDENT_ACTIONS:
            variants = [
                variant for variant in actions[action] if configuration.id in variant.configurations
            ]
            default_id = defaults.get(action)
            if not variants:
                if default_id is not None:
                    _fail(
                        manifest_path,
                        f"Config {configuration.id!r} defaults.{action} references "
                        f"{default_id!r}, but no {action} variants support that Config",
                    )
                continue
            if default_id is None:
                _fail(
                    manifest_path,
                    f"Config {configuration.id!r} must declare defaults.{action} because "
                    f"compatible {action} variants exist",
                )
            if not any(variant.id == default_id for variant in variants):
                _fail(
                    manifest_path,
                    f"Config {configuration.id!r} defaults.{action} {default_id!r} is not a "
                    f"compatible {action} variant",
                )

    for platform in SUPPORTED_REPO_COMMAND_PLATFORMS:
        compatible = [
            configuration
            for configuration in configurations
            if configuration.platforms is None or platform in configuration.platforms
        ]
        if not compatible:
            continue
        default_count = sum(configuration.default for configuration in compatible)
        if default_count != 1:
            _fail(
                manifest_path,
                f"platform {platform} must have exactly one default Config; found {default_count}",
            )


def _required_string(value: object, manifest_path: Path, field_path: str) -> str:
    if not isinstance(value, str) or not value.strip():
        _fail(manifest_path, f"{field_path} must be a non-empty string")
    return value


def _required_string_array(
    value: object,
    manifest_path: Path,
    field_path: str,
) -> tuple[str, ...]:
    if not isinstance(value, list) or not all(isinstance(entry, str) for entry in value):
        _fail(manifest_path, f"{field_path} must be a string array")
    return tuple(value)


def _required_non_empty_string_array(
    value: object,
    manifest_path: Path,
    field_path: str,
) -> tuple[str, ...]:
    result = _required_string_array(value, manifest_path, field_path)
    if not result or any(not entry.strip() for entry in result):
        _fail(manifest_path, f"{field_path} must be a non-empty string array")
    return result


def _optional_string_array(
    value: object,
    manifest_path: Path,
    field_path: str,
) -> tuple[str, ...]:
    if value is _MISSING:
        return ()
    result = _required_string_array(value, manifest_path, field_path)
    if any(not entry.strip() for entry in result):
        _fail(manifest_path, f"{field_path} entries must be non-empty strings")
    return result


def _assert_unique_strings(
    values: Sequence[str],
    manifest_path: Path,
    field_path: str,
    *,
    noun: str = "value",
) -> None:
    seen: set[str] = set()
    for value in values:
        if value in seen:
            _fail(manifest_path, f"{field_path} contains duplicate {noun} {value!r}")
        seen.add(value)


def _assert_repo_relative_path(value: str, manifest_path: Path, field_path: str) -> None:
    normalized = value.replace("\\", "/")
    posix_path = PurePosixPath(normalized)
    if posix_path.is_absolute() or PureWindowsPath(value).is_absolute() or ".." in posix_path.parts:
        _fail(
            manifest_path,
            f"{field_path} path {value!r} must stay within the repository",
        )


def _fail(manifest_path: Path, message: str) -> NoReturn:
    raise RepoCommandManifestError(f"Invalid command manifest {manifest_path}: {message}")
